kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
-      -- information.
-      deferred :: !(Seq (Pair (TermLike variable)))
-    , -- | Matching solution: Substitutions for target variables.
-      substitution :: !(Map (SomeVariableName variable) (TermLike variable))
-    , -- | Matching solution: Additional constraints.
-      predicate :: !(MultiAnd (Predicate variable))
-    , -- | Bound variable that must not escape in the solution.
-      bound :: !(Set (SomeVariableName variable))
-    , -- | Target variables that may be substituted.
-      targets :: !(Set (SomeVariableName variable))
-    , -- | Variables that must not be shadowed.
-      avoiding :: !(Set (SomeVariableName variable))
-    }
-    deriving stock (GHC.Generic)
-
-type MatcherT variable simplifier =
-    RWST (SideCondition variable) () (MatcherState variable) simplifier
-
--- | Pop the next constraint from the matching queue.
-pop ::
-    MonadState (MatcherState variable) matcher =>
-    matcher (Maybe (Pair (TermLike variable)))
-pop = do
-    queued <- Lens.use (field @"queued")
-    case Set.minView queued of
-        Just (next, queued') -> do
-            field @"queued" .= queued'
-            return . Just . getConstraint $ next
-        _ ->
-            return Nothing
-
--- | Push a new constraint onto the matching queue.
-push ::
-    Ord variable =>
-    MonadState (MatcherState variable) matcher =>
-    Pair (TermLike variable) ->
-    matcher ()
-push pair = field @"queued" %= Set.insert (Constraint pair)
-
-{- | Defer a constraint until more information is available.
-
-The constraint will be retried after the next substitution which affects it.
--}
-defer ::
-    MonadState (MatcherState variable) matcher =>
-    Pair (TermLike variable) ->
-    matcher ()
-defer pair = field @"deferred" %= (:|> pair)
-
-{- | Record an element substitution in the matching solution.
+    unwrapSetValue ::
+        [(Value NormalizedSet, Value NormalizedSet)] ->
+        [(TermLike RewritingVariableName, TermLike RewritingVariableName)]
+    unwrapSetValue _ = []
+
+    unwrapMapValue ::
+        [(Value NormalizedMap, Value NormalizedMap)] ->
+        [(TermLike RewritingVariableName, TermLike RewritingVariableName)]
+    unwrapMapValue vals = map (Bifunctor.bimap getMapValue getMapValue) vals
+
+    unwrapSetElement ::
+        Element NormalizedSet ->
+        Element NormalizedSet ->
+        [(TermLike RewritingVariableName, TermLike RewritingVariableName)]
+    unwrapSetElement elem1 elem2 = [(getSetElement elem1, getSetElement elem2)]
+
+    unwrapMapElement ::
+        Element NormalizedMap ->
+        Element NormalizedMap ->
+        [(TermLike RewritingVariableName, TermLike RewritingVariableName)]
+    unwrapMapElement elem1 elem2 =
+        let (key1, val1) = getMapElement elem1
+            (key2, val2) = getMapElement elem2
+         in [(key1, key2), (val1, val2)]
+patternMatch' sideCondition ((MatchItem pat subject boundVars boundSet) : rest) deferred predicate subst = do
```

Comment:
Please refactor `patternMatch'` so that there's only one main body of the function. That's where you should case split on the arguments and call different worker functions which should be named suggestively, according to the job they do.

Right now this function is very hard to read because it spans over 500 LOC and the top level part of it is scattered throughout the file.

Reply:
This is subject to the same issue I mentioned above; putting everything under a single binding would mean that the functions in the where clause would no longer be able to pattern match on fields within the MatchItem being analyzed, which would make the code even harder to read. If you know how to avoid that issue, I'm game, though.